File: sdks/redlite-java/src/main/java/com/redlite/SetOptions.java
package com.redlite;

import org.jetbrains.annotations.Nullable;

/**
 * Options for SET command.
 */
public class SetOptions {
    private final @Nullable Long ex;
    private final @Nullable Long px;
    private final boolean nx;
    private final boolean xx;

    private SetOptions(@Nullable Long ex, @Nullable Long px, boolean nx, boolean xx) {
        this.ex = ex;
        this.px = px;
        this.nx = nx;
        this.xx = xx;
    }

    /**
     * Get expiration time in seconds.
     */
    public @Nullable Long getEx() {
        return ex;
    }

    /**
     * Get expiration time in milliseconds.
     */
    public @Nullable Long getPx() {
        return px;
    }

    /**
     * Only set if key does not exist.
     */
    public boolean isNx() {
        return nx;
    }

    /**
     * Only set if key exists.
     */
    public boolean isXx() {
        return xx;
    }

    /**
     * Create a new builder.
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Builder for SetOptions.
     */
    public static class Builder {
        private Long ex = null;
        private Long px = null;
        private boolean nx = false;
        private boolean xx = false;

        /**
         * Set expiration time in seconds.
         */
        public Builder ex(long seconds) {
            this.ex = seconds;
            return this;
        }

        /**
         * Set expiration time in milliseconds.
         */
        public Builder px(long milliseconds) {
            this.px = milliseconds;
            return this;
        }

        /**
         * Only set if key does not exist.
         */
        public Builder nx() {
            this.nx = true;
            return this;
        }

        /**
         * Only set if key exists.
         */
        public Builder xx() {
            this.xx = true;
            return this;
        }

        /**
         * Build the SetOptions.
         */
        public SetOptions build() {
            return new SetOptions(ex, px, nx, xx);
        }
    }
}

File: sdks/redlite-kotlin/src/main/kotlin/com/redlite/SetOptions.kt
package com.redlite

/**
 * Options for SET command.
 *
 * @property ex Expiration time in seconds
 * @property px Expiration time in milliseconds
 * @property nx Only set if key does not exist
 * @property xx Only set if key exists
 */
data class SetOptions(
    val ex: Long? = null,
    val px: Long? = null,
    val nx: Boolean = false,
    val xx: Boolean = false
) {
    /**
     * Builder pattern for Java interoperability.
     */
    class Builder {
        private var ex: Long? = null
        private var px: Long? = null
        private var nx: Boolean = false
        private var xx: Boolean = false

        fun ex(seconds: Long) = apply { this.ex = seconds }
        fun px(milliseconds: Long) = apply { this.px = milliseconds }
        fun nx() = apply { this.nx = true }
        fun xx() = apply { this.xx = true }
        fun build() = SetOptions(ex, px, nx, xx)
    }

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }
}
